### PR TITLE
CAN Rearchitecture

### PR DESCRIPTION
**Pull request by cindyli-13**
[CAN rearchitecture notes](https://docs.google.com/document/d/15vdVIMkG_4F-1X9uTqsn_lBBEgxPH9hCTVnR2D9pFlY/edit)

[Linked firmware repo PR](https://github.com/uwrobotics/MarsRover2020-firmware/pull/337)

The idea is to use the YAML file to define CAN messages and signals, and generate the rest (DBC file, C headers and source for managing message data structures) with [cantools](https://cantools.readthedocs.io/en/latest/) and python scripts, all based on the YAML description. All auto-generated files are put under `can/generated`. CANMsgMap is a std::unordered_map wrapper (very much like the LookupTable but it's not const), I was hoping this could be shared by FW and SW.

The current CAN dev workflow is:
1. Update the uwrt_mars_rover_can.yaml file
2. Run `python3 scripts/generate_can.py` in the HW bridge repo root
3. Make the necessary app changes in FW or SW, using the yaml file and generated files as reference

### CODE REVIEW

File: include/uwrt_mars_rover_utils/hw_bridge.h [hunk 1]
```diff
@@ -1,14 +1,21 @@
...
+#include "CANMsgMap.h"
+#include "uwrt_mars_rover_can.h"
+#include "uwrt_mars_rover_can_enums.h"
+#include "uwrt_mars_rover_can_wrapper.h"
+
 namespace HWBRIDGE {
+
 namespace ROVERCONFIG {
 constexpr uint16_t ROVER_MOTOR_PWM_FREQ_HZ = 20000;   // 20 kHz
 constexpr uint32_t ROVER_CANBUS_FREQUENCY  = 500000;  // 500 kbps
```
Comment (younesr1) [inline include/uwrt_mars_rover_utils/hw_bridge.h]: forgot to change this

**Review by younesr1 — CHANGES_REQUESTED** [review]
looked through this but it was hard to follow. please be the first to review your code! Id appreciate more detail around the CANMsgMap

File: include/uwrt_mars_rover_utils/hw_bridge.h (hunk 1 above)
Comment (younesr1) [reply include/uwrt_mars_rover_utils/hw_bridge.h]: actually a lot of this section doesnt even belong in the bridge i dont think. some things vars only concern fw

File: can/CANMsgMap.cpp [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+bool CANMsgMap::setSignalValue(CANID msgID, CANSIGNAL signalName, CANSignalValue_t signalValue) {
+  if (contains(msgID, signalName)) {
+    m_map[msgID][signalName] = signalValue;
+    return true;
+  }
+  return false;
+}
+
+bool CANMsgMap::contains(CANID msgID) const {
```
Comment (younesr1) [inline can/CANMsgMap.cpp:25]: since cpp 20, theres a contains functions for maps

File: can/generated/uwrt_mars_rover_can_enums.h [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+
+constexpr uint32_t ROVER_CANBUS_FREQUENCY_HZ = 1000000;
+
+typedef double CANSignalValue_t;
+
+typedef struct {
+  uint8_t raw[8];
+} CANMsgData_t;
+
+enum CANBUSID {
```
Comment (younesr1) [inline can/generated/uwrt_mars_rover_can_enums.h]: prefer enum classes over enums where appropriate

File: can/generated/uwrt_mars_rover_can_enums.h [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+  CANBUS1,
+  CANBUS2,
+};
+
+enum CANID {
+  ARM_SET_CONTROL_MODE              = 0x0,
```
Comment (younesr1) [inline can/generated/uwrt_mars_rover_can_enums.h]: ensure no conflicts with roboteq canids

File: can/generated/uwrt_mars_rover_can_enums.h (hunk 3 above)
Comment (wmmc88) [reply can/generated/uwrt_mars_rover_can_enums.h]: I think this is autogenerated? So changing this would mean needing to fork the upstream repo for the generator tool and making changes there

Comment (cindyli-13) [reply can/generated/uwrt_mars_rover_can_enums.h]: this one's custom auto-generated, I can change it in the python script to make them enum classes

File: can/CANMsgMap.cpp [hunk 5]
```diff
@@ -0,0 +1,55 @@
+#include "CANMsgMap.h"
```
Comment (cindyli-13) [inline can/CANMsgMap.cpp:1]: The structure of a CANMsgMap is:
```
{
msg1Name : {signal1Name : signal1Value, signal2Name : signal2Value, ...},
msg2Name : {signal1Name : signal1Value, signal2Name : signal2Value, ...},
...
}
```

The idea is that each target will have a RX msg map and a TX msg map (for streamed messages only, one-shot messages don't go into the maps). On the RX side, when the CAN driver receives a message intended for this target, it unpacks the message contents and puts the signals into the RX msg map. The application can read from the msg map (through a CAN lib API call) to obtain the most recent value of a RX signal. On the TX side, whenever the application wants to update a TX signal, it writes it to the TX msg map (also through a CAN lib API call). The CAN driver periodically packs the TX messages from the msg map and transmits them.

At least this is how I'm doing it on the FW side, but hoping this CANMsgMap is used on the SW side too since the custom auto-generated CAN functions in the hw bridge are dependent on the CANMsgMap.

File: can/CANMsgMap.h [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+#include <unordered_map>
+
+#include "uwrt_mars_rover_can_enums.h"
+
+namespace HWBRIDGE {
+
+class CANMsgMap {
+ private:
+  // https://stackoverflow.com/questions/18837857/cant-use-enum-class-as-unordered-map-key
+  struct enumHasher {
```
Comment (cindyli-13) [inline can/CANMsgMap.h]: Need a custom key hasher for enum types

File: can/CANMsgMap.h [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+  // copy constructor and assignment
+  CANMsgMap(const CANMsgMap &) = default;
+  CANMsgMap &operator=(const CANMsgMap &) = default;
+
+  // move constructor and assignment
+  CANMsgMap(CANMsgMap &&) = default;
+  CANMsgMap &operator=(CANMsgMap &&) = default;
+
+  // initializer list constructor
+  CANMsgMap(std::initializer_list<std::pair<const CANID, CANSignalMapType>> init);
```
Comment (cindyli-13) [inline can/CANMsgMap.h:29]: Most of this was based on LookupTable in the fw repo

File: can/generated/uwrt_mars_rover_can.c [hunk 8]
```diff
@@ -0,0 +1,2671 @@
+/**
```
Comment (cindyli-13) [inline can/generated/uwrt_mars_rover_can.c:1]: This file is cantools generated, it takes can of packing / unpacking CAN signals into message structs or raw data bytes

File: can/generated/uwrt_mars_rover_can.h [hunk 9]
```diff
@@ -0,0 +1,5288 @@
+/**
```
Comment (cindyli-13) [inline can/generated/uwrt_mars_rover_can.h:1]: This file is cantools generated, it takes can of packing / unpacking CAN signals into message structs or raw data bytes

File: can/generated/uwrt_mars_rover_can_dump.txt [hunk 10]
```diff
@@ -0,0 +1,1513 @@
+================================= Messages =================================
```
Comment (cindyli-13) [inline can/generated/uwrt_mars_rover_can_dump.txt:1]: This file is cantools generated, it's essentially a human-readable dump of the DBC file content. I think it presents the CAN data really well

File: can/generated/uwrt_mars_rover_can_enums.h [hunk 11]
```diff
@@ -0,0 +1,162 @@
+// AUTO-GENERATED FILE. DO NOT MODIFY
```
Comment (cindyli-13) [inline can/generated/uwrt_mars_rover_can_enums.h]: This file is custom auto-generated (not affiliated with cantools). It wraps the cantools generated macros into enums. I'm also gonna put signal choice enums here later

File: can/generated/uwrt_mars_rover_can_wrapper.cpp [hunk 12]
```diff
@@ -0,0 +1,2975 @@
+// AUTO-GENERATED FILE. DO NOT MODIFY
```
Comment (cindyli-13) [inline can/generated/uwrt_mars_rover_can_wrapper.cpp]: This file is custom auto-generated (not affiliated with cantools). It's a wrapper for packing / unpacking signals into or from CANMsgMaps using the cantools-generated functions.

File: include/uwrt_mars_rover_utils/hw_bridge.h (hunk 1 above)
Comment (cindyli-13) [reply include/uwrt_mars_rover_utils/hw_bridge.h]: Sorry I should've left a comment earlier, this hw_bridge,h is still under construction. Eventually all the CAN stuff will be moved to a separate header. The actual ROVER_CANBUS_FREQUENCY is under can/generated/uwrt_mars_rover_can_enums.h

File: scripts/generate_can.py [hunk 13]
```diff
@@ -0,0 +1,247 @@
+#!/usr/bin/env python3
```
Comment (cindyli-13) [inline scripts/generate_can.py:1]: This is the script that gets called for generating all the auto-generated CAN files. So after making a change to the YAML CAN file, you'd call `python3 scripts/generate_can.py` from the hw bridge root (using python3 cuz cantools isn't supported by python2) and it'll re-generate all the files under can/generated (and also update some parts of the YAML). 

This script reads from the YAML and uses the cantools API to create the DBC file, then invokes a cmd line cantools call to generate the C source and header from the DBC. The script then calls the generate_can_enums.py and generate_can_wrapper.py scripts to generate the uwrt_mars_rover_can_enums.h, uwrt_mars_rover_can_wrapper.h, and uwrt_mars_rover_can_wrapper.cpp files.

File: scripts/generate_can.py [hunk 14]
```diff
@@ -0,0 +1,247 @@
...
+YAML_FILE_PATH = os.path.join(CAN_FOLDER_PATH, YAML_FILE_NAME)
+
+
+AUTOGEN_message_enums = {}
+AUTOGEN_signal_enums = []
+AUTOGEN_msg_map = {}
+AUTOGEN_original_msg_names = {}
+
+
+# https://github.com/eerimoq/cantools/blob/871581b57785fdbd79c118878f9b9c148984963c/cantools/database/can/c_source.py#L762
```
Comment (cindyli-13) [inline scripts/generate_can.py:44]: I haven't migrated to using a cantools git submodule yet, so just using the cantools python package right now. This camel_to_snake_case() conversion was copied from their source code

**Comment by cindyli-13** [comment]
> looked through this but it was hard to follow. please be the first to review your code! Id appreciate more detail around the CANMsgMap

sorry my bad, some parts of this PR are still a WIP, but I just left some comments to explain what each of the files do

File: can/generated/uwrt_mars_rover_can_enums.h (hunk 4 above)
Comment (cindyli-13) [reply can/generated/uwrt_mars_rover_can_enums.h]: yep CAN IDs are still TBD rn, will update them soon

File: include/uwrt_mars_rover_utils/hw_bridge.h [hunk 15]
```diff
@@ -1,108 +1,19 @@
...
 
 namespace ARM {
 enum class ActuatorID { TURNTABLE, SHOULDER, ELBOW, WRISTLEFT, WRISTRIGHT, CLAW };
 }
 
 namespace CONTROL {
-enum class Mode { OPEN_LOOP, VELOCITY, POSITION, CURRENT };
+enum class Mode { OPEN_LOOP, POSITION, VELOCITY, CURRENT };
 
 namespace SAFETY {
```
Comment (cindyli-13) [inline include/uwrt_mars_rover_utils/hw_bridge.h]: This safety section will eventually be replaced by CAN enums

File: scripts/generate_can.py [hunk 16]
```diff
@@ -0,0 +1,258 @@
...
+        # else if only given scale/offset (no min/max)
+        elif 'scale' in signal and 'offset' in signal:
+            scale = signal['scale']
+            offset = signal['offset']
+
+            # calculate min/max based on scale/offset and length
+            sig_min = - (2**(length-1)-1) * scale + \
+                offset if is_signed else offset
+            sig_max = (2**(length-1)-1) * scale + \
+                offset if is_signed else (2**length-2) * scale + offset
```
Comment (cindyli-13) [inline scripts/generate_can.py:130]: This conversion calculation from min/max to scale/offset makes it so that the range does not include the SNA value (currently set to largest possible value representable by the bits). I realized that a caveat of this is that the `signal_in_range()` check will fail if the signal value is SNA. Does this approach make sense?

File: can/uwrt_mars_rover_can.yaml [hunk 17]
```diff
@@ -0,0 +1,1855 @@
+uwrt_mars_rover_can_bus:
+  bus_frequency: 1000000
+  canid_filter_mask: 2016
+  canid_filters:
+    0: NO_FILTER
```
Comment (cindyli-13) [inline can/uwrt_mars_rover_can.yaml]: CAN IDs are represented in decimal in the yaml. May be worth looking into how to have hex representation instead

File: can/CANMsgMap.h (hunk 6 above)
Comment (wmmc88) [reply can/CANMsgMap.h]: According to your linked post, this isn't needed anymore and was corrected in c++ 14. 
https://stackoverflow.com/a/29618545/10173605

File: can/generated/uwrt_mars_rover_can_dump.txt (hunk 10 above)
Comment (wmmc88) [reply can/generated/uwrt_mars_rover_can_dump.txt:1]: ohh i like this. will make changes easier to review for those not familiar with dbc structure

File: scripts/generate_can.py (hunk 14 above)
Comment (wmmc88) [reply scripts/generate_can.py:44]: i think the version of cantools needs to be version controlled. submoduled fork if you're modifying it or requirements.txt if you're just using the package. the generated code can change from version to version and we need to make sure to have a concrete way of selecting which version

**Review by wmmc88 — COMMENTED** [review]
It'd be nice if the auto-generated file's comments that say its autogenerated specify what actually generated it. So that future devs looking at generated files can see what tools generated which files

File: can/CANMsgMap.cpp (hunk 5 above)
Comment (cindyli-13) [reply can/CANMsgMap.cpp:1]: Edit: changed it so that one-shot RX messages are also put in the CANMsgsMaps. This just offloads the message unpacking logic to the CAN library instead of the app, to make the application code more readable. For TX one-shots, it's still the responsibility of the app to package up the message into raw data form before handing it to the CAN library

**Comment by cindyli-13** [comment]
Going to merge this PR so that we can continue development with the new CAN stuff
